### ISSUE TITLE
HTTP Basic Authentication wrong decoding

### ISSUE THREAD

**Issue by GdotBdot** [mitreid-connect/OpenID-Connect-Java-Spring-Server#857]
OAuth requires the client id to be encoded "using the 'application/x-www-form-urlencoded' encoding algorithm" (http://tools.ietf.org/html/rfc6749#section-2.3.1). This is not respected when a OIDC-Client accesses the UserInfo Endpoint in the Authorization Code Flow: If the client id contains e.g. an "@", the missing decoding results in an "invalid_client" error response.
A quick look at the code suggests that Spring's built-in ```DaoAuthenticationProvider``` is used to decode the Basic Auth headers, and it's retrieving the client data from the ```DefaultOAuth2ClientDetailsEntityService```. The former was not explicitly written for use in OAuth, and the latter does not know the requested client id comes from an HTTP Basic Auth.

**Comment by jricher** [comment]
The client doesn't send its client ID when using the UserInfo endpoint -- it should only be sending the access token at that point. If you mean the token endpoint, then this could be a bug. Are you seeing the issue on the client or the server side? Which branch/version?

**Comment by GdotBdot** [comment]
You are right, it's the token endpoint.
I use the current code on master. I tried to connect using the example client of connect2id (https://bitbucket.org/connect2id/openid-connect-dev-client), with the client id "@1234".

An ```InvalidClientException``` is thrown in ```DefaultOAuth2ClientDetailsEntityService.loadClientByClientId``` on line 207 (and catched elsewhere, I got this from debugging), and the server returns a 401 Response
```
Error code: invalid_client
Error description: Client with id %401234 was not found
Error URI: null
```
The error description is the message from the said exception.

The connect2id client does the encoding correctly, it's also discussed here: https://bitbucket.org/connect2id/oauth-2.0-sdk-with-openid-connect-extensions/issue/129/clientsecretbasic-incorrectly-url-encodes
